### PR TITLE
map works with both functions and keys as the second argument

### PR DESCRIPTION
**Pull request by phoniks**

### CODE REVIEW

**Comment by phoniks** [comment]
@jrob8577

File: spec/map_spec.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+var users = [
+  { 'user': 'barney' },
+  { 'user': 'fred' }
+];
+
+
+describe('intersection', ()=> {
+  it('returns the intersection of values present in all input arrays', () => {
+    expect(map([1, 2, 3, 4], (n) => {return n * n})).toEqual([1, 4, 9, 16])
```
Comment (jrobcodes) [inline spec/map_spec.js]: ES6 allows us to make this more concise:

``` es6
expect( map( [1, 2, 3, 4], n = > n * n ) ).toEqual( [1, 4, 9, 16 ] )
```

Recall that in ES6, a function that takes only one argument may omit the parentheses, and a function that has only a single statement in the body that is the return statement may omit the curly braces.

File: spec/map_spec.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { map } from '../src/map'
+
+var users = [
+  { 'user': 'barney' },
+  { 'user': 'fred' }
+];
+
```
Comment (jrobcodes) [inline spec/map_spec.js:7]: Remove one of these new lines.

File: spec/map_spec.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+describe('intersection', ()=> {
```
Comment (jrobcodes) [inline spec/map_spec.js]: Is this testing `map` or `intersection`?

**Comment by jrobcodes** [comment]
@phoniks The tests are confusing - I'm guessing you forgot to update a copy and paste from another file to make this meaningful for map?  I would also add some tests to handle edge cases (empty arrays and null, for example).  You are missing tests around functionality related to the iteratee - I think we need tests that verify that if an iteratee is not passed, the identity function is used, as well as tests that ensure that iteratees are invoked with the necessary arguments (value, index, array for an array input, and value, key, object for an object input).

File: src/map.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+const map = ( array, operation ) => {
+  const fn = operation || ( () => {} )
+  let results =[]
+  if(typeof operation === 'function'){
```
Comment (jrobcodes) [inline src/map.js]: For clarity, I would separate these two different map approaches into their own functions, and then `map` could look something like:

``` es6
return {
  'function': mapWithFunction,
  'string': mapWithString
}[ typeof operation ]( array, fn )

// Elsewhere:
const mapWithFunction = (array, operation) => {
  // etc
}
```

File: src/map.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    for(let index = 0; index < array.length; index++ ){
+      let item = array[index]
```
Comment (jrobcodes) [inline src/map.js:8]: I would omit all of these temporary variables, in favor of:

``` es6
results.push( fn( array[ index ] ) )
```

File: src/map.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+      let result = fn(item)
+      results.push(result)
+    }
+  }
+  else if (typeof operation === 'string'){
+    for(let index = 0; index < array.length; index++ ){
+      let item = array[index]
+      let result = item[operation]
+      results.push(result)
+      console.log(operation)
```
Comment (jrobcodes) [inline src/map.js]: Please remove all `console.log`s before issuing PR

File: src/map.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    for(let index = 0; index < array.length; index++ ){
+      let item = array[index]
+      let result = fn(item)
+      results.push(result)
+    }
+  }
+  else if (typeof operation === 'string'){
+    for(let index = 0; index < array.length; index++ ){
+      let item = array[index]
+      let result = item[operation]
```
Comment (jrobcodes) [inline src/map.js]: Given that this is the only difference between the two different approaches, I would look at ways to simplify this code.  Let's discuss; probably easier to do in person than via comments.  :P

Comment (phoniks) [inline src/map.js]:
``` es6
let result = typeof( operation ) === 'function' ? fn( item ) : item[ operation ]
```
